app/models/katello/concerns/host_managed_extensions.rb

```diff
@@ -122,6 +128,43 @@ def import_enabled_repositories(repos)
         content_facet.update_repositories_by_paths(paths.compact)
       end
 
+      def import_module_streams(module_streams)
+        # TODO: Implement this when module stream support is added
+        streams = {}
+        module_streams.each do |module_stream|
+          stream = AvailableModuleStream.where(name: module_stream["name"],
+                                               stream: module_stream["stream"]).first_or_create!
+          streams[stream.id] = module_stream
+        end
+        sync_available_module_stream_associations(streams)
+      end
+
+      def sync_available_module_stream_associations(new_available_module_streams)
+        updatable_streams = self.host_available_module_streams.where(:available_module_stream_id => new_available_module_streams.keys)
+
+        old_associated_ids = self.available_module_stream_ids
+
+        delete_ids = old_associated_ids - new_available_module_streams.keys
+        if delete_ids.any?
+          self.host_available_module_streams.where(:available_module_stream_id => delete_ids).delete_all
+        end
+        new_ids = new_available_module_streams.keys - old_associated_ids
+        new_ids.each do |new_id|
+          module_stream = new_available_module_streams[new_id]
+          self.host_available_module_streams.create!(host_id: self.id,
+                                          available_module_stream_id: new_id,
+                                          installed_profiles: module_stream["installed_profiles"],
+                                          status: module_stream["status"])
+        end
+
+        updatable_streams.each do |hams|
+          module_stream = new_available_module_streams[hams.available_module_stream_id]
+          if module_stream["status"] != hams.status || module_stream["installed_profiles"] != hams.installed_profiles
```

Comment:
could you check `unless module_stream == hams` or is there a reason to specifically check those fields?

Reply:
continuing that thought, could you use `first_or_create!` for the `hams` object?

Reply:
Oh may be the way this code is written is confusing all I am trying to in sync_available_module_stream_associations is
1. find HAMS associations to delete
1. find HAMS associations to create (new associations)
1. find HAMS associations to update

So the way I approach each step given current hams_ids and new hams_ids 
1. delete current - new
1. create new - current
1. update current 'intersection' new. Now I say "update only if status changed or installed profiles changed" because those are the only fields that are subject to changing.

Reply:
@parthaa How do you feel about something like this that uses the actual intersection?

```ruby
        updatable_streams.each do |hams|
          module_stream = new_available_module_streams[hams.available_module_stream_id]
          shared_keys = hams.attributes.keys & module_stream.keys
          module_stream_data = module_stream.slice(*shared_keys)
          if hams.attributes.slice(*shared_keys) != module_stream_data
            hams.update_attributes!(module_stream_data)
          end
        end
```

Reply:
Forgive me, I'm going to be a bit nitpicky here..

Can you separate the chunks of logic with spaces? i.e.
```ruby
        updatable_streams = self.host_available_module_streams.where(:available_module_stream_id => new_available_module_streams.keys)
        old_associated_ids = self.available_module_stream_ids
        delete_ids = old_associated_ids - new_available_module_streams.keys

        if delete_ids.any?
          self.host_available_module_streams.where(:available_module_stream_id => delete_ids).delete_all
        end

        new_ids = new_available_module_streams.keys - old_associated_ids
        new_ids.each do |new_id|
          module_stream = new_available_module_streams[new_id]
          self.host_available_module_streams.create!(host_id: self.id,
                                                     available_module_stream_id: new_id,
                                                     installed_profiles: module_stream["installed_profiles"],
                                                     status: module_stream["status"])
        end

        updatable_streams.each do |hams|
          module_stream = new_available_module_streams[hams.available_module_stream_id]
          shared_keys = hams.attributes.keys & module_stream.keys
          module_stream_data = module_stream.slice(*shared_keys)
          if hams.attributes.slice(*shared_keys) != module_stream_data
            hams.update_attributes!(module_stream_data)
          end
        end
```

Reply:
Updated!